In [1]:
import pandas as pd

In [2]:
data=pd.read_csv('news.csv')

In [3]:
data

Unnamed: 0                                              title  \
0           8476                       You Can Smell Hillary’s Fear   
1          10294  Watch The Exact Moment Paul Ryan Committed Pol...   
2           3608        Kerry to go to Paris in gesture of sympathy   
3          10142  Bernie supporters on Twitter erupt in anger ag...   
4            875   The Battle of New York: Why This Primary Matters   
...          ...                                                ...   
6330        4490  State Department says it can't find emails fro...   
6331        8062  The ‘P’ in PBS Should Stand for ‘Plutocratic’ ...   
6332        8622  Anti-Trump Protesters Are Tools of the Oligarc...   
6333        4021  In Ethiopia, Obama seeks progress on peace, se...   
6334        4330  Jeb Bush Is Suddenly Attacking Trump. Here's W...   

                                                   text label  
0     Daniel Greenfield, a Shillman Journalism Fello...  FAKE  
1     Google Pinterest Digg Linkedin Reddit Stumbleu...  FAKE  
2     U.S. Secretary of State John F. Kerry said Mon...  REAL  
3     — Kaydee King (@KaydeeKing) November 9, 2016 T...  FAKE  
4     It's primary day in New York and front-runners...  REAL  
...                                                 ...   ...  
6330  The State Department told the Republican Natio...  REAL  
6331  The ‘P’ in PBS Should Stand for ‘Plutocratic’ ...  FAKE  
6332   Anti-Trump Protesters Are Tools of the Oligar...  FAKE  
6333  ADDIS ABABA, Ethiopia —President Obama convene...  REAL  
6334  Jeb Bush Is Suddenly Attacking Trump. Here's W...  REAL  

[6335 rows x 4 columns]

In [4]:
dataset=data.drop("Unnamed: 0",axis=1)

In [5]:
dataset

title  \
0                          You Can Smell Hillary’s Fear   
1     Watch The Exact Moment Paul Ryan Committed Pol...   
2           Kerry to go to Paris in gesture of sympathy   
3     Bernie supporters on Twitter erupt in anger ag...   
4      The Battle of New York: Why This Primary Matters   
...                                                 ...   
6330  State Department says it can't find emails fro...   
6331  The ‘P’ in PBS Should Stand for ‘Plutocratic’ ...   
6332  Anti-Trump Protesters Are Tools of the Oligarc...   
6333  In Ethiopia, Obama seeks progress on peace, se...   
6334  Jeb Bush Is Suddenly Attacking Trump. Here's W...   

                                                   text label  
0     Daniel Greenfield, a Shillman Journalism Fello...  FAKE  
1     Google Pinterest Digg Linkedin Reddit Stumbleu...  FAKE  
2     U.S. Secretary of State John F. Kerry said Mon...  REAL  
3     — Kaydee King (@KaydeeKing) November 9, 2016 T...  FAKE  
4     It's primary day in New York and front-runners...  REAL  
...                                                 ...   ...  
6330  The State Department told the Republican Natio...  REAL  
6331  The ‘P’ in PBS Should Stand for ‘Plutocratic’ ...  FAKE  
6332   Anti-Trump Protesters Are Tools of the Oligar...  FAKE  
6333  ADDIS ABABA, Ethiopia —President Obama convene...  REAL  
6334  Jeb Bush Is Suddenly Attacking Trump. Here's W...  REAL  

[6335 rows x 3 columns]

In [6]:
x=dataset['text']
y=dataset['label']

In [7]:
from sklearn.model_selection import train_test_split

In [8]:
from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(
    x, y, test_size=0.2, random_state=42
)


In [9]:
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer, HashingVectorizer

In [10]:
count_vectorizer = CountVectorizer(stop_words='english')
count_train = count_vectorizer.fit_transform(x_train)
print(count_train)
count_test = count_vectorizer.transform(x_test)

<Compressed Sparse Row sparse matrix of dtype 'int64'
	with 1321067 stored elements and shape (5068, 61359)>
  Coords	Values
  (0, 17037)	1
  (0, 55902)	1
  (0, 44697)	1
  (0, 30414)	1
  (0, 18744)	1
  (0, 27192)	2
  (0, 41513)	1
  (0, 48710)	1
  (0, 29397)	1
  (0, 48889)	1
  (0, 52337)	1
  (0, 43055)	1
  (0, 12501)	1
  (0, 45993)	1
  (1, 17037)	1
  (1, 55902)	4
  (1, 37651)	1
  (1, 38054)	2
  (1, 710)	2
  (1, 1143)	1
  (1, 41683)	2
  (1, 41828)	1
  (1, 23788)	1
  (1, 12034)	1
  (1, 40627)	1
  :	:
  (5067, 4549)	1
  (5067, 25010)	1
  (5067, 57115)	1
  (5067, 57822)	1
  (5067, 58825)	1
  (5067, 42325)	1
  (5067, 41502)	1
  (5067, 12750)	1
  (5067, 9822)	1
  (5067, 18733)	2
  (5067, 25217)	1
  (5067, 56806)	1
  (5067, 60241)	1
  (5067, 16223)	1
  (5067, 10309)	4
  (5067, 26511)	1
  (5067, 18781)	1
  (5067, 11011)	3
  (5067, 36419)	1
  (5067, 26926)	4
  (5067, 58826)	1
  (5067, 33325)	1
  (5067, 18112)	1
  (5067, 57662)	1
  (5067, 18512)	1


In [11]:
count_train.shape


(5068, 61359)

In [12]:
from sklearn.naive_bayes import MultinomialNB
from sklearn import metrics

In [13]:
clf = MultinomialNB()

clf.fit(count_train, y_train)
pred = clf.predict(count_test)
score = metrics.accuracy_score(y_test, pred)
print("accuracy:   %0.3f" % score)

cm = metrics.confusion_matrix(y_test, pred, labels=['FAKE', 'REAL'])

accuracy:   0.899


In [14]:
from sklearn.metrics import classification_report

report=classification_report(y_test, pred)

In [15]:
print(report)

              precision    recall  f1-score   support

        FAKE       0.93      0.87      0.89       628
        REAL       0.88      0.93      0.90       639

    accuracy                           0.90      1267
   macro avg       0.90      0.90      0.90      1267
weighted avg       0.90      0.90      0.90      1267



In [16]:
from sklearn.linear_model import LogisticRegression, Perceptron
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.naive_bayes import GaussianNB, MultinomialNB, BernoulliNB
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier, ExtraTreesClassifier, GradientBoostingClassifier, AdaBoostClassifier
from sklearn.svm import SVC, LinearSVC
from sklearn.neural_network import MLPClassifier
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report


In [17]:
models = [
    # Linear models
    ("Logistic Regression", LogisticRegression(max_iter=1000)),
       # Naive Bayes
    ("Gaussian NB", GaussianNB()),
    ("Multinomial NB", MultinomialNB()),
    ("Bernoulli NB", BernoulliNB()),

    # Distance based
    ("KNN", KNeighborsClassifier(n_neighbors=5)),

    # Tree based
    ("Decision Tree", DecisionTreeClassifier()),
      ("Random Forest", RandomForestClassifier(n_estimators=100)),

    # Boosting
    ("AdaBoost", AdaBoostClassifier()),
    ("Gradient Boosting", GradientBoostingClassifier()),

    # SVM
    ("Linear SVM", LinearSVC()),
    ("Kernel SVM", SVC(kernel='rbf')),

   ]


In [18]:
for name, model in models:
    try:
        model.fit(count_train, y_train)
        y_pred = model.predict(count_test)
        acc = accuracy_score(y_test, y_pred)
        print(f"{name}: {acc:.3f}")
    except Exception as e:
        print(f"{name}: Not suitable ({e})")


Logistic Regression: 0.927
Gaussian NB: Not suitable (Sparse data was passed for X, but dense data is required. Use '.toarray()' to convert to a dense numpy array.)
Multinomial NB: 0.899
Bernoulli NB: 0.829
KNN: 0.809
Decision Tree: 0.817
Random Forest: 0.905
AdaBoost: 0.848
Gradient Boosting: 0.883


C:\Users\anand\anaconda3\Lib\site-packages\sklearn\svm\_base.py:1249: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(


Linear SVM: 0.890
Kernel SVM: 0.879


In [19]:
from sklearn.feature_extraction.text import TfidfVectorizer

tfidf = TfidfVectorizer(stop_words='english')

X_train_tfidf = tfidf.fit_transform(x_train)
X_test_tfidf  = tfidf.transform(x_test)


In [20]:
for name, model in models:
    try:
        model.fit(X_train_tfidf, y_train)
        y_pred = model.predict(X_test_tfidf)
        acc = accuracy_score(y_test, y_pred)
        print(f"{name}: {acc:.3f}")
    except Exception as e:
        print(f"{name}: Not suitable ({e})")

Logistic Regression: 0.916
Gaussian NB: Not suitable (Sparse data was passed for X, but dense data is required. Use '.toarray()' to convert to a dense numpy array.)
Multinomial NB: 0.845
Bernoulli NB: 0.829
KNN: 0.562
Decision Tree: 0.814
Random Forest: 0.908
AdaBoost: 0.848
Gradient Boosting: 0.895
Linear SVM: 0.936
Kernel SVM: 0.931


In [22]:
from sklearn.feature_extraction.text import HashingVectorizer

hash_vec = HashingVectorizer(
    n_features=2**20,
    stop_words='english',
    alternate_sign=False
)

X_train_hash = hash_vec.transform(x_train)
X_test_hash  = hash_vec.transform(x_test)


In [23]:
for name, model in models:
    try:
        model.fit(count_train, y_train)
        y_pred = model.predict(count_test)
        acc = accuracy_score(y_test, y_pred)
        print(f"{name}: {acc:.3f}")
    except Exception as e:
        print(f"{name}: Not suitable ({e})")


Logistic Regression: 0.927
Gaussian NB: Not suitable (Sparse data was passed for X, but dense data is required. Use '.toarray()' to convert to a dense numpy array.)
Multinomial NB: 0.899
Bernoulli NB: 0.829
KNN: 0.809
Decision Tree: 0.811
Random Forest: 0.910
AdaBoost: 0.848
Gradient Boosting: 0.883


C:\Users\anand\anaconda3\Lib\site-packages\sklearn\svm\_base.py:1249: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(


Linear SVM: 0.890
Kernel SVM: 0.879


# Summary
### for CountVectorizer , Logistic regression algorithm gives a better accuracy wihich is 0.92
### for TF-IDF , Linear SVMClassifier algorithm gives a better accuracy wihich is 0.93
### for HashingVectorizer, Logistic regression algorithm gives a better accuracy wihich is 0.92